src/main/java/io/reactivex/internal/operators/flowable/FlowableGroupBy.java

```diff
@@ -420,14 +430,22 @@ public boolean isEmpty() {
     static final class EvictionAction<K, V> implements Consumer<GroupedUnicast<K,V>> {
 
         final Queue<GroupedUnicast<K, V>> evictedGroups;
+        final AtomicBoolean cancelled;
 
-        EvictionAction(Queue<GroupedUnicast<K, V>> evictedGroups) {
+        EvictionAction(Queue<GroupedUnicast<K, V>> evictedGroups, AtomicBoolean cancelled) {
             this.evictedGroups = evictedGroups;
+            this.cancelled = cancelled;
         }
 
         @Override
-        public void accept(GroupedUnicast<K,V> value) {
-            evictedGroups.offer(value);
+        public void accept(GroupedUnicast<K,V> group) {
+            if (!cancelled.get()) {
+                evictedGroups.offer(group);
+            } else {
+                // note that this call to `onComplete` is safe because 
+                // GroupedUnicast is designed to allow concurrent access
+                group.onComplete();
```

Comment:
I'm concerned about a race with `group.onNext` where cancelled has just happened and `group.onComplete` gets called above which then drains in `State`. At that point `group.onNext` gets called from another thread and that emission is lost. I'm guessing some significant rework might be needed to protect against this. Thoughts?